### DISCUSSION TITLE
Better way to get the location state?

### DISCUSSION THREAD

**Discussion by geohuz** [remix-run/remix#5465] [Q&A] [answered]
I have a login form in my home page, and have a action function to submit the login credential, on successful login I would like to redirect to a url which user tried before login, or if there is no previous url then always redirect to a fixed router:

```jsx
let location
export async function action({request}) {
    const formData = await request.formData()
    const loginData = Object.fromEntries(formData)
    const auth = queryClient.rootStore.auth
    await auth.login(loginData)
    if (auth.token!=="none") {
      // the redirection logic
      return redirect(location.state ? location.state.from.pathname : "/contacts")
    } else {
      return loginData
    }
}

function Home() {
  // assign to the global location which awkward
  location = useLocation()
}

```

As I can't use hook inside action, I have to assign the locaiton variable in the functional component: Home, and then use it inside action function, which is awkward. Is there anyway to deal with the issue?

**Comment by sergiodxa** [comment, marked as answer]
Assigning to a module variable the value of useLocation will not work because the action runs server-side and on a new request.

If the action is on the same route of the component then you don't really need the location, because you know where the user was before, on the same route!

If the action is on another route, then you can use one of these methods to redirect the user back to where it was https://sergiodxa.com/articles/redirect-to-the-original-url-inside-a-remix-action.
